programs/vote/src/vote_state/mod.rs

```diff
@@ -188,18 +188,35 @@ fn check_update_vote_state_slots_are_valid(
     }
 
     // Check if the proposed root is too old
-    if let Some(new_proposed_root) = vote_state_update.root {
-        // If the root is less than the earliest slot hash in the history such that we
-        // cannot verify whether the slot was actually was on this fork, set the root
-        // to the current vote state root for safety.
+    let original_proposed_root = vote_state_update.root;
+    if let Some(new_proposed_root) = original_proposed_root {
+        // If the new proposed root `R` is less than the earliest slot hash in the history
+        // such that we cannot verify whether the slot was actually was on this fork, set
+        // the root to the latest vote in the current vote that's less than R.
         if earliest_slot_hash_in_history > new_proposed_root {
             vote_state_update.root = vote_state.root_slot;
+            let mut prev_slot = Slot::MAX;
+            let current_root = vote_state_update.root;
+            for lockout in vote_state.votes.iter().rev() {
+                let is_bigger_than_root = current_root
+                    .map(|current_root| lockout.slot > current_root)
+                    .unwrap_or(true);
+                // Ensure we're iterating from biggest to smallest vote in the
+                // current vote state
+                assert!(lockout.slot < prev_slot && is_bigger_than_root);
+                if lockout.slot <= new_proposed_root {
+                    vote_state_update.root = Some(lockout.slot);
+                    break;
+                }
+                prev_slot = lockout.slot;
+            }
         }
     }
 
-    // index into the new proposed vote state's slots, starting with the root if it exists then
-    // we use this mutable root to fold the root slot case into this loop for performance
-    let mut check_root = vote_state_update.root;
+    // Index into the new proposed vote state's slots, starting with the root if it exists then
+    // we use this mutable root to fold checking the root slot into this the below loop
```

Comment:
-this : `we use this mutable root to fold checking the root slot into the below loop"`